lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,630 @@
+      if (element is FunctionElement ||
+          element.enclosingElement3 is ExtensionElement) {
+        parentElement = element.enclosingElement3!;
+      } else if (node.parent is CascadeExpression) {
+        // TODO: handle cascade notation
+        return;
+      } else if (node.target == null) {
+        return;
+      } else {
+        // TODO: determining parentElement through node.target seems unreliable...
+        parentElement = node.target!.staticType?.element2! ??
+            (node.target! as Identifier).staticElement!;
+      }
+      identifyDependencyName();
+    } on AnalysisException {
+      // do not continue if this invocation is unfit for analysis
+      return;
+    }
+
+    late bool constraintIssue;
+
+    // differentiate between class methods and top-level functions
+    if (parentElement is ClassElement) {
+      // ignore generics
+      if ((parentElement as ClassElement).typeParameters.isNotEmpty) {
+        return;
+      }
+      constraintIssue = !dependencyShape.containsMethodWithName(
+          parentElement.name!, symbolName);
+    } else if (parentElement is CompilationUnitElement) {
+      constraintIssue = !dependencyShape.containsFunctionWithName(symbolName);
+    } else if (parentElement is ExtensionElement) {
+      constraintIssue = !dependencyShape.containsExtensionMethodWithName(
+          parentElement.name!, symbolName);
+    } else {
+      // context.warning('Subclass ${parentElement.toString()} of parentElement (method/function) is not supported.');
+      return;
+    }
+
+    issues[elementId] = constraintIssue
+        ? LowerBoundConstraintIssue(
+      dependencyPackageName: packageName,
+            constraint: context.dependencies[packageName]!.version,
+            currentVersion: context.findInstalledVersion(packageName),
+            lowestVersion: Version.parse(dependencyShape.version),
+            identifier: symbolName,
+            parentName: element.kind == ElementKind.FUNCTION
+                ? null
+                : parentElement.name!,
+            kind: element.kind,
+            parentKind: element.kind == ElementKind.FUNCTION
+                ? null
+                : parentElement.kind,
+            references: [span],
+          )
+        : null;
+  }
+}
+
+/// Returns the package name from a library identifier/uri, or null if
+/// [libraryUri] isn't of the form `package:*/*`.
+///
+/// Example:
+/// `packageFromLibraryUri('package:my_package/my_library.dart')` returns
+/// `'my_package'`.
+String? packageFromLibraryUri(String libraryUri) {
+  if (libraryUri.startsWith('package:') && libraryUri.contains('/')) {
+    final endIndex = libraryUri.indexOf('/');
+    return libraryUri.substring(8, endIndex);
+  }
+  return null;
+}
+
+// TODO: rename this class? it is still useful to have this structure even if we are not sure whether this is a real issue or a false positive - consider LowerBoundConstraintIssueResult or similar
+class LowerBoundConstraintIssue {
```

Comment:
rename this class? it is still useful to have this structure even if we are not sure whether this is a _real_ issue or a _false positive_ - consider `LowerBoundConstraintIssueResult` or similar

Reply:
PotentialLowerBoundConstraintIssue